### PR TITLE
Add Multi-Index Search

### LINKED ISSUE #415: Implement multi-search method
In the [v1.1.0 of Meilisearch](https://github.com/meilisearch/meilisearch/releases/tag/v1.1.0), the [multi-search](https://docs.meilisearch.com/reference/api/multi_search.html) was introduced.

This feature is not yet but should be implemented as per the [expectation of the integrations](https://github.com/meilisearch/integration-guides/issues/251).

- See related issue in `meilisearch` https://github.com/meilisearch/meilisearch/issues/3427
- See specification https://github.com/meilisearch/specifications/pull/225
- See `Perform multiple searches in one single HTTP request` section in the integration implementation requirements: https://github.com/meilisearch/integration-guides/issues/251

### PR DESCRIPTION
**Pull request by ahmednfwela**
# Pull Request

## Related issue
Fixes #415

## What does this PR do?
- Introduces `ISearchableJsonConverterFactory`, `ISearchableJsonConverter<T>`
- Add `IndexUid` to `ISearchable<T>`
- Introduce `MultiSearchQuery`, `MultiSearchResult`
- Add `Task<MultiSearchResult> MultiSearchAsync(MultiSearchQuery)` to `MeilisearchClient`

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

cc @alallema after their vacation :)

### CODE REVIEW

**Comment by bidoubiwa** [comment]
Hey @ahmednfwela,

Thanks so much for this valuable contribution ✨ Currently, the maintainer of this repository @alallema is on holiday so they'll not be able to review you. They'll be back next week!

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs:44]: Why did you have to cast here?

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs]:
```suggestion
```

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs]: I think you can use the entire line here instead of breaking into new lines!

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs:61]: I didn't know about this way of casting, very cool!

File: src/Meilisearch/MeilisearchClient.cs
Comment (brunoocasali) [inline src/Meilisearch/MeilisearchClient.cs:96]:
```suggestion

            var responseMessage = await _http.PostAsJsonAsync("multi-search", query, Constants.JsonSerializerOptionsRemoveNulls, cancellationToken: cancellationToken);
```

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs]:
```suggestion
    public class ISearchableJsonConverter<T> : JsonConverter<ISearchable<T>> where T : class
```

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs:30]:
```suggestion
    }

```

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (brunoocasali) [inline src/Meilisearch/ISearchableJsonConverter.cs]:
```suggestion
        public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
```

**Review by brunoocasali — COMMENTED** [review]
Hi @ahmednfwela amazing work here! you rocks! :D 

I'll let @alallema handle that PR because she is the owner of the repo ;)

**Comment by alallema** [comment]
Hi @ahmednfwela,
Thank you very much for this PR, I will review it tomorrow at the latest. Sorry for the delay

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (ahmednfwela) [reply src/Meilisearch/ISearchableJsonConverter.cs:44]: because of 
image

File: src/Meilisearch/ISearchableJsonConverter.cs
Comment (ahmednfwela) [reply src/Meilisearch/ISearchableJsonConverter.cs]: I am using https://csharpier.com/ which breaks long lines automatically sorry about that 😅

File: src/Meilisearch/MultiSearchResult.cs
Comment (alallema) [inline src/Meilisearch/MultiSearchResult.cs:16]: I'm pretty sure it's not doable, but why can't we do something like:
```
public class MultiSearchResult<T>
...
public List<ISearchable<T>> Results { get; set; }
```
For after the fact, just do something like:
```
            foreach (var item in result.Results)
            {
                if (item is PaginatedSearchResult<Movie> paginated) {
                    /// ... Do something with your movie ...
                }
                if (item is SearchResult<Movie> normal) {
                    /// ... Do something with your movie ...
                }
            }
```

File: src/Meilisearch/PaginatedSearchResult.cs
Comment (alallema) [inline src/Meilisearch/PaginatedSearchResult.cs:101]: This is much better than writing the description twice, as is currently the case! Can you do it for every field?
If it's too annoying, I can open an issue about it 😊

File: src/Meilisearch/SearchResult.cs
Comment (alallema) [inline src/Meilisearch/SearchResult.cs:86]: Same than for the first one

**Review by alallema — COMMENTED** [review]
Hi @ahmednfwela,
Me again, really sorry for the delay and thank you very much this RP is 🔥 🔥 🔥. 
Just added a question or two, but this PR could be merged like that 🚀

File: src/Meilisearch/MultiSearchResult.cs
Comment (ahmednfwela) [reply src/Meilisearch/MultiSearchResult.cs:16]: the problem is that c# doesn't support type unions, and the main target of multi index search is to search across multiple indexes which might be of different types

Comment (ahmednfwela) [reply src/Meilisearch/MultiSearchResult.cs:16]: we can of course go down the rabbit hole of `OneOf<T1,T2,...>` but that would add more unnecessary complexity

File: src/Meilisearch/PaginatedSearchResult.cs
Comment (ahmednfwela) [reply src/Meilisearch/PaginatedSearchResult.cs:101]: sure! I was planning on doing that anyway 😄

Comment (ahmednfwela) [reply src/Meilisearch/PaginatedSearchResult.cs:101]: https://github.com/meilisearch/meilisearch-dotnet/pull/416/commits/dc6c9b6c8769b2e297a43fa91ab6bdc23119a8df

File: src/Meilisearch/SearchResult.cs
Comment (ahmednfwela) [reply src/Meilisearch/SearchResult.cs:86]: https://github.com/meilisearch/meilisearch-dotnet/pull/416/commits/dc6c9b6c8769b2e297a43fa91ab6bdc23119a8df

File: src/Meilisearch/MultiSearchResult.cs
Comment (alallema) [reply src/Meilisearch/MultiSearchResult.cs:16]: Thank you very much for all these explanations I think indeed that it is better as you have already implemented 😊

**Review by alallema — DISMISSED** [review]
LGTM! 🚀 ✨ 
Thank you again for this contribution @ahmednfwela and for contributing to Meilisearch ❤️

File: tests/Meilisearch.Tests/MultiIndexSearchTests.cs
Comment (brunoocasali) [inline tests/Meilisearch.Tests/MultiIndexSearchTests.cs]: Can you add a test here asserting that you can access fields like `genre`, and `title`?

**Review by brunoocasali — COMMENTED** [review]
Just adding this use case, I think it will be nice for test regressions.

File: tests/Meilisearch.Tests/MultiIndexSearchTests.cs
Comment (brunoocasali) [reply tests/Meilisearch.Tests/MultiIndexSearchTests.cs]: After it I think we are good to go!

Comment (ahmednfwela) [reply tests/Meilisearch.Tests/MultiIndexSearchTests.cs]: good catch! I forgot to use the proper Json serialization options in the example
https://github.com/meilisearch/meilisearch-dotnet/pull/416/commits/dc1affcf7a88d7c37b7d24722680e1d96e8e44de

**Review by alallema — APPROVED** [review]

**Review by brunoocasali — APPROVED** [review]
Thank you @ahmednfwela 👏 👏 

It's your call now @alallema 🪖

**Comment by alallema** [comment]
Bors merge

**Pull request merged**
